test/e2e_node/criproxy_test.go

```diff
@@ -0,0 +1,151 @@
+//go:build linux
+// +build linux
+
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2enode
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	kubeletevents "k8s.io/kubernetes/pkg/kubelet/events"
+	"k8s.io/kubernetes/test/e2e/feature"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	"k8s.io/kubernetes/test/e2e_node/criproxy"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+// Examples of using CRI proxy
+var _ = SIGDescribe(feature.CriProxy, framework.WithSerial(), func() {
+	f := framework.NewDefaultFramework("cri-proxy-example")
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+
+	ginkgo.Context("Inject a pull image error exception into the CriProxy", func() {
+		ginkgo.AfterEach(func() {
+			resetCRIProxyInjector()
+		})
+
+		ginkgo.It("Pod failed to start due to an image pull error.", func(ctx context.Context) {
+			expectedErr := fmt.Errorf("PullImage failed")
+			addCRIProxyInjector(func(apiName string) error {
+				if apiName == criproxy.PullImage {
+					return expectedErr
+				}
+				return nil
+			})
+
+			pod := e2epod.NewPodClient(f).Create(ctx, newPullImageAlwaysPod())
+			podErr := e2epod.WaitForPodRunningInNamespace(ctx, f.ClientSet, pod)
+			gomega.Expect(podErr).To(gomega.HaveOccurred())
+
+			eventMsg, err := getFailedToPullImageMsg(ctx, f, pod.Name)
+			framework.ExpectNoError(err)
+			isExpectedErrMsg := strings.Contains(eventMsg, expectedErr.Error())
```

Comment:
errors.New() and errors.Is() should be used instead of strings.Contains()

Reply:
The expected content of eventMsg is `Failed to pull image \"registry.k8s.io/pause:3.10\": PullImage failed`. Do you mean to create a new error with eventMsg? I feel that the current implementation is simpler."

Reply:
My initial idea of using strings.Contains to determine if the error messages are equal isn't the best approach, because it only checks if part of the error message exists, without considering the context or other possible nuances. However, since here we are determining whether the error wrapped by the kubelet or the event matches the expected error, let's keep it as is.